docs: Organize gadget devel documentation (#3208)
inspektor-gadget/inspektor-gadget: docs/gadget-devel/gadget-intro.md

---
title: Gadget Introduction
sidebar_position: 100
description: Introduction to Gadgets
---

Gadgets are the central component in the Inspektor Gadget framework. A Gadget is
an [OCI image](https://opencontainers.org/) that includes one or more eBPF
programs, metadata YAML file and, optionally, WASM modules for post processing,

> etc. As OCI images, they use the same tooling as containers and share the same

Reviewer: "etc."? What else can be there? I would remove it.

Author: Ideas so far:
- Logo
- Documentation (Readme file)
- Source code
- Tests
 
I know we aren't using them now, but I'm sure we'll have more things there soon.

> programs, metadata YAML file and, optionally, WASM modules for post processing,
> etc. As OCI images, they use the same tooling as containers and share the same
> attributes; shareable, modular, deployable, etc.

Reviewer: What do you mean with tooling? You mean you can use `oras`, `crane`, etc. to manipulate them? I think it's confusing because then you talk about "deployment", so I don't want users to think that they can run Gadgets with docker 😅. 

```suggestion
programs, metadata YAML file and, optionally, WASM modules for post processing.
As OCI images, they can be stored in a container registry (compliant with the
OCI specifications), making them easy to distribute and share.
```

Author: I took this from #3221, would you mind making the comment there? I'll align it later on.

> attributes; shareable, modular, deployable, etc.

Author: Ok, that change is originally coming from #3221 as well. Let's get all those comments there and I'll update it later on. (I'm already taking to @blixtra about it)

collected and presented.

### Tracers

Tracers are data sources that provide a stream of events as they happen on the
system: a file is opened, a DNS request is performed, etc. These data sources use
a [perf ring
buffer](https://docs.kernel.org/next/userspace-api/perf_ring_buffer.html) or
[BPF ring buffer](https://docs.kernel.org/6.6/bpf/ringbuf.html) to transfer the
events from the Gadget to Inspektor Gadget.

Gadgets providing this kind of data source need to define an event `struct` with
the fields they provide:

```c
struct event {
	// all fields here
};
```

Tracers are marked by using the `GADGET_TRACER()` macro. It accepts the following parameters:

```c
GADGET_TRACER(name, mapname, structname);
```

- name: Name of the data source
- mapname: Name of the perf ring buffer of BPF ring buffer used by the program
- structname: Name of the structure defining the gadget's event

Examples of Gadgets that use this kind of data sources are
[trace_open](https://github.com/inspektor-gadget/inspektor-gadget/tree/main/gadgets/trace_open),
[trace_exec](https://github.com/inspektor-gadget/inspektor-gadget/tree/main/gadgets/trace_exec),
etc.

### Map Iterators

> Map Iterators are use to report statistics like number of files being opened,

Reviewer: ```suggestion
Map Iterators are used to report statistics like number of files being opened,
```

bytes going through a network connection, etc. The information is saved by the
Gadget on hash maps where it's then read by Inspektor Gadget.

> Gadgets need to define the types for the key and value and a hash map use to store the information:

Reviewer: ```suggestion
Gadgets need to define the types for the key and value and a hash map used to store the information:
```
